File: AlgorithmsPackPythonSolutions/Parallel Courses III.py
class Solution:
    def minimumTime(self, n: int, relations: List[List[int]], time: List[int]) -> int:
        def addEdge(graph, from_node, to):
            graph[from_node].append(to)

        def topSort(graph, time):
            nodes = len(graph)
            incomingEdges = [0] * nodes

            for i in range(nodes):
                for j in graph[i]:
                    incomingEdges[j] += 1

            q = deque()
            for i in range(nodes):
                if incomingEdges[i] == 0:
                    q.append(i)

            result = [0] * nodes
            while q:
                cur = q.popleft()
                result[cur] += time[cur]

                for j in graph[cur]:
                    incomingEdges[j] -= 1
                    if incomingEdges[j] == 0:
                        q.append(j)
                    result[j] = max(result[j], result[cur])

            return result

        graph = [[] for _ in range(n)]
        for rel in relations:
            addEdge(graph, rel[0] - 1, rel[1] - 1)

        timeComp = topSort(graph, time)
        res = max(timeComp)
        return res

File: AlgorithmsPackCPP/Parallel Courses III.cpp
class Solution {
    typedef vector<vector<int>>GRAPH;

    void addEdge(GRAPH&graph, int from, int to){
        graph[from].push_back(to);
    }
public:
    vector<int>topSort(GRAPH&graph, vector<int>&time){
        int nodes = graph.size();
        vector<int>incomingEdges(nodes);
        for(int i = 0; i < nodes; i++){
            for(int j:graph[i]){
                incomingEdges[j]++;
            }
        }
        queue<int>q;
        for(int i = 0; i < nodes; i++){
            if(incomingEdges[i]==0){
                q.push(i);
            }
        }
        vector<int>result(nodes);
        while(!q.empty()){
            int cur = q.front();
            q.pop();
            result[cur]+=time[cur];

            for(int j:graph[cur]){
                if(--incomingEdges[j]==0){
                    q.push(j);
                }
                result[j] = max(result[j],result[cur]);
            }
        }
        return result;
    }

    int minimumTime(int n, vector<vector<int>>& relations, vector<int>& time) {
        GRAPH graph(n);
        for(auto&rel:relations)
            addEdge(graph,rel[0]-1,rel[1]-1);

        vector<int>timeComp = topSort(graph,time);
        int res = *max_element(timeComp.begin(),timeComp.end());
        return res;
    }
};
